lib/cluster_holding.rb

```diff
@@ -24,4 +24,41 @@ def move(new_cluster)
     end
   end
 
+  # Updates a matching holding or adds it
+  def update
+    c = Cluster.find_by(ocns: @holding.ocn)
+    return cluster unless c
+
+    old_holding = c.holdings.to_a.find do |h|
+      h == @holding &&
+        h.date_received != @holding.date_received
+    end
+    if old_holding
+      old_holding.update_attributes(date_received: @holding.date_received)
+      c
+    else
+      cluster
+    end
+  end
+
+  # Removes a holding
+  def delete
+    c = Cluster.find_by(ocns: @holding.ocn)
+    @holding.delete
+    c.save
+    c.reload
+    c.delete unless c._children.any?
+  end
```

Comment:
The delete should only happen if we deleted the last child, so I'm not sure how a race condition would happen. Happy to entertain a sequence of events that could cause one, though.

Reply:
Hmm, that could happen, although I agree it's very unlikely. I think there is an atomic mongodb `delete_if` operation that might be helpful here? I'm also not sure how `_children` works, so I'm not sure this will ever actually return true - would the holding ever be the only thing in `_children`? I also wonder if we have the same race condition concern for all the split/merge operations - let's say you were in the middle of doing a merge operation that caused one cluster to be deleted, but another thread already had a reference to that to-be-deleted cluster. Even if you could lock that cluster for updates (and I need to check whether that's possible in MongoDB) I think what would happen is that the cluster would get deleted but then in the other thread save would resurrect it? Maybe there's a way to have save raise an exception if you try to save something that got deleted in the meantime, and then it could just retry that operation? That seems like a potentially frequent concern, in any case. Will have to investigate.

Reply:
are we worried about a race condition here?

Always, but I'm still working on an understanding of multithreading with Mongoid. https://docs.mongodb.com/mongoid/current/tutorials/mongoid-transactions/

._children are a document's embedded documents. In the case of a Cluster, that's anything other than the ocns field. I suspect clusters will frequently consist of a single ocn and the holding record it is taken from since we have 400m of them. Looking at it again, this functionally identical to Reclusterer.new(c).recluster